File: packages/shared-types/src/provenance.ts
/** Mirrors `app/schemas/provenance.py`. */

export interface ProvenanceAction {
  action: string | null;
  when: string | null;
  software_agent: string | null;
  parameters?: unknown;
}

export interface ProvenanceValidationFailure {
  code: string;
  explanation: string | null;
}

export interface NormalizedProvenance {
  engine: string;
  engine_version: string;
  has_c2pa: boolean;
  /** null when no manifest exists (nothing to validate). */
  valid_signature: boolean | null;
  signer: string | null;
  signature_alg: string | null;
  signed_at: string | null;
  claim_generator: string | null;
  title: string | null;
  active_manifest: string | null;
  manifest_count: number;
  ingredient_count: number;
  assertion_labels: string[];
  actions: ProvenanceAction[];
  authors: string[];
  validation_codes: string[];
  validation_failures: ProvenanceValidationFailure[];
  warnings: string[];
  /** Structured validation view: engine state, per-family codes, per-ingredient results. */
  validation: {
    state?: string | null;
    source?: string;
    active_manifest?: {
      success: string[];
      informational: string[];
      failure: string[];
    };
    ingredients?: Record<
      string,
      { success: string[]; informational: string[]; failure: string[] }
    >;
  };
  /** `--info` facts reported by the engine (manifest store size, count, validated). */
  info: Record<string, unknown>;
  /** Signed declarations read from the active manifest (T045). */
  assertions: ProvenanceAssertions;
  software_agents: ProvenanceSoftwareAgent[];
  /** Ingredient tree with per-ingredient validation codes. */
  ingredients: ProvenanceIngredient[];
  ingredient_failures: number;
  /** Every manifest in the store, active first, with signing times. */
  manifest_chain: ProvenanceManifestLink[];
  manifest_order_conflict: boolean;
  /** Certificate subject common name (0.28+ engines). */
  signer_common_name: string | null;
  /** embedded | remote | sidecar | none | unknown */
  manifest_location: string;
  /** Host of a remote manifest the engine did not fetch (fetching is disabled). */
  remote_manifest_host: string | null;
}

export interface ProvenanceHashCoverage {
  label: string;
  alg: string | null;
  name: string | null;
  exclusion_count: number;
  exclusions: Array<{ start: number; length: number | null }>;
}

export interface ProvenanceSourceType {
  action: string | null;
  uri: string;
  short: string;
}

export interface ProvenanceIdentity {
  present: boolean;
  kind: string | null;
  names: string[];
  referenced_assertions?: number;
}

export interface ProvenanceAssertions {
  hash_data?: ProvenanceHashCoverage;
  source_types?: ProvenanceSourceType[];
  training_mining?: Record<string, string>;
  identity?: ProvenanceIdentity;
}

export interface ProvenanceSoftwareAgent {
  name: string;
  version: string | null;
  origin: string;
}

export interface ProvenanceIngredient {
  title: string | null;
  format: string | null;
  relationship: string | null;
  document_id: string | null;
  instance_id: string | null;
  manifest_label: string | null;
  validation_codes: string[];
  failure_codes: string[];
  thumbnail_identifier: string | null;
  children: ProvenanceIngredient[];
}

export interface ProvenanceManifestLink {
  label: string;
  parent: string | null;
  signed_at: string | null;
  claim_generator: string | null;
  signer: string | null;
  signed_after_parent?: boolean;
}

export interface ImageProvenanceResponse {
  normalized: NormalizedProvenance;
  manifests: Record<string, unknown>;
  validation_status: Array<Record<string, unknown>>;
  limitations: string[];
  /** `c2patool --tree` text diagram, when produced. */
  tree: string | null;
}

File: services/api/app/schemas/provenance.py
from typing import Any

from pydantic import BaseModel


class ProvenanceActionResponse(BaseModel):
    action: str | None
    when: str | None
    software_agent: str | None
    parameters: Any | None = None


class ProvenanceValidationFailure(BaseModel):
    code: str
    explanation: str | None


class NormalizedProvenanceResponse(BaseModel):
    engine: str
    engine_version: str
    has_c2pa: bool
    valid_signature: bool | None
    signer: str | None
    signature_alg: str | None
    signed_at: str | None
    claim_generator: str | None
    title: str | None
    active_manifest: str | None
    manifest_count: int
    ingredient_count: int
    assertion_labels: list[str]
    actions: list[ProvenanceActionResponse]
    authors: list[str]
    validation_codes: list[str]
    validation_failures: list[ProvenanceValidationFailure]
    warnings: list[str]
    # Structured validation view (state, per-family codes for the active manifest and
    # per-ingredient results) and `--info` facts; empty for rows written before T044.
    validation: dict[str, Any] = {}
    info: dict[str, Any] = {}
    # T045: signed declarations, software agents, ingredient tree and manifest chain.
    assertions: dict[str, Any] = {}
    software_agents: list[dict[str, Any]] = []
    ingredients: list[dict[str, Any]] = []
    ingredient_failures: int = 0
    manifest_chain: list[dict[str, Any]] = []
    manifest_order_conflict: bool = False


class ImageProvenanceResponse(BaseModel):
    normalized: NormalizedProvenanceResponse
    manifests: dict[str, Any]
    validation_status: list[dict[str, Any]]
    limitations: list[str]
    # `c2patool --tree` text diagram of the manifest store, when the engine produced one.
    tree: str | None = None
